File: nRF Toolbox/UART/UARTViewController.m
/*
 * Copyright (c) 2015, Nordic Semiconductor
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "UARTViewController.h"
#import "SWRevealViewController.h"
#import "ScannerViewController.h"
#import "Constants.h"
#import "LogViewController.h"
#import "AppUtilities.h"
#import "EditPopupViewController.h"
#import "LogViewController.h"

@interface UARTViewController ()

@property (strong, nonatomic) BluetoothManager *bluetoothManager;

@property (strong, nonatomic) NSString *uartPeripheralName;
@property (strong, nonatomic) NSMutableArray *buttonsCommands;
@property (strong, nonatomic) NSMutableArray *buttonsHiddenStatus;
@property (strong, nonatomic) NSMutableArray *buttonsImageNames;

@property (strong, nonatomic) NSArray *buttonIcons;
@property (strong, nonatomic) UIButton *selectedButton;

@property (weak, nonatomic) IBOutlet UIButton *connectButton;
@property (weak, nonatomic) IBOutlet UILabel *deviceName;
@property (weak, nonatomic) IBOutlet UILabel *verticalLabel;
@property (weak, nonatomic) IBOutlet UIButton *editButton;
@property (strong, nonatomic) IBOutletCollection(UIButton) NSArray* buttons;

@property (nonatomic, weak) LogViewController* logger;
@property (nonatomic) BOOL editMode;

- (IBAction)editButtonPressed:(UIButton *)sender;
- (IBAction)connectOrDisconnectClicked:(UIButton *)sender;
- (IBAction)showLogButtonClidked:(id)sender;

@end

@implementation UARTViewController

@synthesize connectButton;
@synthesize deviceName;
@synthesize logger;

-(id)initWithCoder:(NSCoder *)decoder
{
    self = [super initWithCoder:decoder];
    if (self) {
        // Custom initialization
        self.buttonIcons = @[@"Stop",@"Play",@"Pause",@"FastForward",@"Rewind",@"End",@"Start",@"Shuffle",@"Record",@"Number_1",
                             @"Number_2",@"Number_3",@"Number_4",@"Number_5",@"Number_6",@"Number_7",@"Number_8",@"Number_9",];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Rotate the vertical label
    self.verticalLabel.transform = CGAffineTransformRotate(CGAffineTransformMakeTranslation(-20.0f, 0.0f), (float)(-M_PI / 2));
    
    // Retrieve three arrays (icons names (NSString), commands (NSString), visibility(Bool)) from NSUserDefaults
    [self retrieveButtonsConfigurations];
    self.editMode = NO;
    
    // Configure the SWRevealViewController
    SWRevealViewController *revealViewController = self.revealViewController;
    if (revealViewController)
    {
        [self.view addGestureRecognizer:self.revealViewController.panGestureRecognizer];
        logger = (LogViewController*) revealViewController.rearViewController;
    }
}

- (IBAction)showLogButtonClidked:(id)sender
{
    [self.revealViewController revealToggleAnimated:YES];
}

- (IBAction)editButtonPressed:(UIButton *)sender
{
    self.editMode = !self.editMode;
}

#pragma mark - Navigation

-(BOOL)shouldPerformSegueWithIdentifier:(NSString *)identifier sender:(id)sender
{
    // The 'scan' seque will be performed only if bluetoothManager == nil (if we are not connected already).
    return ![identifier isEqualToString:@"scan"] || self.bluetoothManager == nil;
}

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"scan"])
    {
        // Set this contoller as scanner delegate
        UINavigationController *nc = segue.destinationViewController;
        ScannerViewController *controller = (ScannerViewController *)nc.childViewControllerForStatusBarHidden;
        controller.delegate = self;
    }
}

#pragma mark - Scanner Delegate methods

-(void)centralManager:(CBCentralManager *)manager didPeripheralSelected:(CBPeripheral *)peripheral
{
    // We may not use more than one Central Manager instance. Let's just take the one returned from Scanner View Controller
    self.bluetoothManager = [[BluetoothManager alloc] initWithManager: manager];
    self.bluetoothManager.delegate = self;
    self.bluetoothManager.logger = logger;
    
    [self.bluetoothManager connectDevice:peripheral];
}

-(void)appDidEnterBackground:(NSNotification *)_notification
{
    [AppUtilities showBackgroundNotification:[NSString stringWithFormat:@"You are still connected to %@ peripheral.",self.uartPeripheralName]];
}

-(void)appDidEnterForeground:(NSNotification *)_notification
{
    [[UIApplication sharedApplication] cancelAllLocalNotifications];
}

- (IBAction)connectOrDisconnectClicked:(UIButton *)sender {
    [self.bluetoothManager disconnectDevice];
}

#pragma mark - UART API

-(void)send:(NSString *)value
{
    if (self.bluetoothManager)
    {
        [self.bluetoothManager send:value];
    }
}

#pragma mark - Buttons behaviour

-(void)setEditMode:(BOOL)editMode
{
    _editMode = editMode;
    
    if (editMode)
    {
        [self.editButton setTitle:@"Done" forState:UIControlStateNormal];
        for (UIButton *button in self.buttons) {
            [button setBackgroundColor:[UIColor colorWithRed:222.0f/255.0f green:74.0f/255.0f blue:19.0f/255.0f alpha:1.0f]];
            [button setEnabled:YES];
        }
    }
    else
    {
        [self.editButton setTitle:@"Edit" forState:UIControlStateNormal];
        [self showButtonsWithSavedConfiguration];
    }
}

-(void)retrieveButtonsConfigurations
{
    NSUserDefaults *buttonsConfigurations = [NSUserDefaults standardUserDefaults];
    
    if ([buttonsConfigurations objectForKey:@"buttonsCommands"]) //Buttons configurations already saved in NSUserDefaults
    {
        // Retrieving the saved values
        self.buttonsCommands = [NSMutableArray arrayWithArray:[buttonsConfigurations objectForKey:@"buttonsCommands"]];
        self.buttonsHiddenStatus = [NSMutableArray arrayWithArray:[buttonsConfigurations objectForKey:@"buttonsHiddenStatus"]];
        self.buttonsImageNames = [NSMutableArray arrayWithArray:[buttonsConfigurations objectForKey:@"buttonsImageNames"]];
        [self showButtonsWithSavedConfiguration];
    }
    else //First time viewcontroller is loaded and there is no saved buttons configurations in NSUserDefaults
    {
        // Setting up the default values for the first time
        self.buttonsCommands = [[NSMutableArray alloc]initWithArray:@[@"",@"",@"",@"",@"",@"",@"",@"",@""]];
        self.buttonsHiddenStatus = [[NSMutableArray alloc]initWithArray:@[@YES,@YES,@YES,@YES,@YES,@YES,@YES,@YES,@YES]];
        self.buttonsImageNames = [[NSMutableArray alloc]initWithArray:@[@"Play",@"Play",@"Play",@"Play",@"Play",@"Play",@"Play",@"Play",@"Play"]];
        
        [buttonsConfigurations setObject:self.buttonsCommands forKey:@"buttonsCommands"];
        [buttonsConfigurations setObject:self.buttonsHiddenStatus forKey:@"buttonsHiddenStatus"];
        [buttonsConfigurations setObject:self.buttonsImageNames forKey:@"buttonsImageNames"];
        [buttonsConfigurations synchronize];
    }
}

-(void)showButtonsWithSavedConfiguration
{
    for (UIButton *button in self.buttons)
    {
        if ([self.buttonsHiddenStatus[[button tag]-1] boolValue])
        {
            [button setBackgroundColor:[UIColor colorWithRed:200.0f/255.0f green:200.0f/255.0f blue:200.0f/255.0f alpha:1.0f]];
            [button setImage:nil forState:UIControlStateNormal];
            [button setEnabled:NO];
        }
        else
        {
            [button setBackgroundColor:[UIColor colorWithRed:0.0f/255.0f green:156.0f/255.0f blue:222.0f/255.0f alpha:1.0f]];
            UIImage *image = [UIImage imageNamed:self.buttonsImageNames[[button tag]-1]];
            [button setImage:image forState:UIControlStateNormal];
            [button setEnabled:YES];
        }
    }
}

#pragma mark - Edit action handling

//One out of 9 Remote Buttons is pressed
- (IBAction)buttonPressed:(id)sender
{
    if (self.editMode)
    {
        self.selectedButton = (UIButton*)sender;
        [self showPopoverOnButton];
    }
    else
    {
        NSString *command = self.buttonsCommands[[sender tag]-1];
        [self send:command];
    }
}

-(void)showPopoverOnButton
{
    EditPopupViewController *popoverVC = [self.storyboard instantiateViewControllerWithIdentifier:@"StoryboardIDEditPopup"];
    popoverVC.delegate = self;
    popoverVC.isHidden = NO; //[self.buttonsHiddenStatus[[self.selectedButton tag]-1 ] boolValue]; // Show it when the popover opens
    popoverVC.command = self.buttonsCommands[[self.selectedButton tag]-1];
    NSString *buttonImageName = self.buttonsImageNames[[self.selectedButton tag] -1];
    popoverVC.iconIndex = (int)[self.buttonIcons indexOfObject:buttonImageName];
    popoverVC.modalPresentationStyle = UIModalPresentationPopover;
    popoverVC.popoverPresentationController.delegate = self;
    [self presentViewController:popoverVC animated:YES completion:nil];
    
    popoverVC.popoverPresentationController.sourceView = self.view;
    popoverVC.popoverPresentationController.sourceRect = self.view.bounds;
    [popoverVC.popoverPresentationController setPermittedArrowDirections:0];
    popoverVC.preferredContentSize = CGSizeMake(300.0, 300.0);
}

//Delegate of UIPopoverPresentationControllerDelegate
-(UIModalPresentationStyle)adaptivePresentationStyleForPresentationController:(UIPresentationController *)controller
{
    return UIModalPresentationNone;
}

//Delegate of ButtonConfigureDelegate is called when user will press OK button on EditPopupViewController
- (void) didButtonConfigured:(NSString*)command iconIndex:(int)index shouldHideButton:(BOOL)status {
    NSUserDefaults *buttonsConfigurations = [NSUserDefaults standardUserDefaults];
    
    int buttonTag = (int)[self.selectedButton tag] - 1;
    
    self.buttonsHiddenStatus[[self.selectedButton tag] - 1] = [NSNumber numberWithBool:status];
    [buttonsConfigurations setObject:self.buttonsHiddenStatus forKey:@"buttonsHiddenStatus"];
    
    if (!status)
    {
        UIImage *image = [UIImage imageNamed:self.buttonIcons[index]];
        [self.selectedButton setImage:image forState:UIControlStateNormal];
    }
    else
    {
        [self.selectedButton setImage:nil forState:UIControlStateNormal];
    }
    
    self.buttonsImageNames[buttonTag] = self.buttonIcons[index];
    [buttonsConfigurations setObject:self.buttonsImageNames forKey:@"buttonsImageNames"];
    
    self.buttonsCommands[buttonTag] = command;
    [buttonsConfigurations setObject:self.buttonsCommands forKey:@"buttonsCommands"];
    
    [buttonsConfigurations synchronize];
}

#pragma mark - BluetoothManager delegate methods

-(void)didDeviceConnected:(NSString *)peripheralName
{
    // Scanner uses other queue to send events. We must edit UI in the main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        logger.bluetoothManager = self.bluetoothManager;
        self.uartPeripheralName = peripheralName;
        [deviceName setText:peripheralName];
        [connectButton setTitle:@"DISCONNECT" forState:UIControlStateNormal];
    });
    
    //Following if condition display user permission alert for background notification
    if ([UIApplication instancesRespondToSelector:@selector(registerUserNotificationSettings:)]) {
        [[UIApplication sharedApplication] registerUserNotificationSettings:[UIUserNotificationSettings settingsForTypes:UIUserNotificationTypeAlert|UIUserNotificationTypeSound categories:nil]];
    }
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(appDidEnterBackground:) name:UIApplicationDidEnterBackgroundNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(appDidEnterForeground:) name:UIApplicationDidBecomeActiveNotification object:nil];
}

-(void)didDeviceDisconnected
{
    // Scanner uses other queue to send events. We must edit UI in the main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        logger.bluetoothManager = nil;
        [connectButton setTitle:@"CONNECT" forState:UIControlStateNormal];
        [deviceName setText:@"DEFAULT UART"];
        
        if ([AppUtilities isApplicationStateInactiveORBackground]) {
            [AppUtilities showBackgroundNotification:[NSString stringWithFormat:@"%@ peripheral is disconnected", self.uartPeripheralName]];
        }
        self.uartPeripheralName = nil;
        
    });
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationDidBecomeActiveNotification object:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationDidEnterBackgroundNotification object:nil];
    
    self.bluetoothManager = nil;
}

-(void)isDeviceReady
{
    NSLog(@"Device is ready");
}

-(void)deviceNotSupported
{
    NSLog(@"Device not supported");
}

@end

File: nRF Toolbox/UART/UARTViewController.swift
//
//  UARTViewController.swift
//  nRF Toolbox
//
//  Created by Mostafa Berg on 11/05/16.
//  Copyright © 2016 Nordic Semiconductor. All rights reserved.
//

import UIKit
import CoreBluetooth

class UARTViewController: UIViewController, BluetoothManagerDelegate, ScannerDelegate, UIPopoverPresentationControllerDelegate, ButtonConfigureDelegate, StoryboardInstantiable {
    
    //MARK: - View Properties
    var bluetoothManager    : BluetoothManager?
    var uartPeripheralName  : String?
    var buttonsCommands     : NSMutableArray?
    var buttonsHiddenStatus : NSMutableArray?
    var buttonsImageNames   : NSMutableArray?
    var buttonIcons         : NSArray
    var selectedButton      : UIButton?
    var logger              : LogViewController?
    var editMode            : Bool!

    //MARK: - View Actions
    @IBAction func connectionButtonTapped(_ sender: AnyObject) {
        bluetoothManager?.cancelPeripheralConnection()
    }
    @IBAction func editButtonTapped(_ sender: AnyObject) {
        let currentEditMode = editMode!
        setEditMode(mode: !currentEditMode)
    }
    @IBAction func showLogButtonTapped(_ sender: AnyObject) {
        revealViewController().revealToggle(animated: true)
    }
    @IBAction func buttonTapped(_ sender: AnyObject){
        if editMode {
            selectedButton = sender as? UIButton;
            showPopoverOnButton()
        } else {
            let command = buttonsCommands![sender.tag-1]
            send(value: String(describing: command))
        }
    }

    //MARK: - View OUtlets
    @IBOutlet weak var verticalLabel: UILabel!
    @IBOutlet weak var editButton: UIButton!
    @IBOutlet weak var deviceName: UILabel!
    @IBOutlet var buttons: [UIButton]!
    @IBOutlet weak var connectionButton: UIButton!
    
    //MARK: - UIViewControllerDelegate
    required init?(coder aDecoder: NSCoder) {
        buttonIcons = ["Stop","Play","Pause","FastForward","Rewind","End","Start","Shuffle","Record","Number_1",
        "Number_2","Number_3","Number_4","Number_5","Number_6","Number_7","Number_8","Number_9",]
        super.init(coder: aDecoder)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Rotate the vertical label
        verticalLabel.transform = CGAffineTransform(translationX: -(verticalLabel.frame.width/2) + (verticalLabel.frame.height / 2), y: 0.0).rotated(by: -.pi / 2);
        
        // Retrieve three arrays (icons names (NSString), commands (NSString), visibility(Bool)) from NSUserDefaults
        retrieveButtonsConfiguration()
        editMode = false
        
        // Configure the SWRevealViewController
        let revealViewController = self.revealViewController()
        if revealViewController != nil {
            view.addGestureRecognizer(revealViewController!.panGestureRecognizer())
            logger = revealViewController?.rearViewController as? LogViewController
        }
    }
    
    //MARK: - Segue methods
    override func shouldPerformSegue(withIdentifier identifier: String, sender: Any?) -> Bool {
        // The 'scan' seque will be performed only if bluetoothManager == nil (if we are not connected already).
        return identifier != "scan" || bluetoothManager == nil
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        guard segue.identifier == "scan" else {
            return
        }
        
        // Set this contoller as scanner delegate
        let nc = segue.destination as! UINavigationController
        let controller = nc.children.first as! ScannerViewController
        // controller.filterUUID = CBUUID.init(string: ServiceIdentifiers.uartServiceUUIDString)
        controller.delegate = self
    }
    
    //MARK: - UIPopoverPresentationCtonrollerDelegate
    func adaptivePresentationStyle(for controller: UIPresentationController) -> UIModalPresentationStyle {
        return UIModalPresentationStyle.none
    }
    
    //MARK: - ScannerViewDelegate
    func centralManagerDidSelectPeripheral(withManager aManager: CBCentralManager, andPeripheral aPeripheral: CBPeripheral) {
        // We may not use more than one Central Manager instance. Let's just take the one returned from Scanner View Controller
        bluetoothManager = BluetoothManager(withManager: aManager)
        bluetoothManager!.delegate = self
        bluetoothManager!.logger = logger
        logger!.clearLog()
        
        if let name = aPeripheral.name {
            uartPeripheralName = name
            deviceName.text = name
        } else {
            uartPeripheralName = "device"
            deviceName.text = "No name"
        }
        connectionButton.setTitle("CANCEL", for: .normal)
        bluetoothManager!.connectPeripheral(peripheral: aPeripheral)
    }
    
    //MARK: - BluetoothManagerDelegate
    func peripheralReady() {
        print("Peripheral is ready")
    }
    
    func peripheralNotSupported() {
        print("Peripheral is not supported")
    }
    
    func didConnectPeripheral(deviceName aName: String?) {
        // Scanner uses other queue to send events. We must edit UI in the main queue
        DispatchQueue.main.async {
            self.logger!.bluetoothManager = self.bluetoothManager
            self.connectionButton.setTitle("DISCONNECT", for: .normal)
        
            //Following if condition display user permission alert for background notification
            if UIApplication.instancesRespond(to: #selector(UIApplication.registerUserNotificationSettings(_:))){
                UIApplication.shared.registerUserNotificationSettings(UIUserNotificationSettings(types: [.sound, .alert], categories: nil))
            }

            NotificationCenter.default.addObserver(self, selector: #selector(self.applicationDidEnterBackgroundCallback), name: UIApplication.didEnterBackgroundNotification, object: nil)
            NotificationCenter.default.addObserver(self, selector: #selector(self.applicationDidBecomeActiveCallback), name: UIApplication.didBecomeActiveNotification, object: nil)
        }
    }

    func didDisconnectPeripheral() {
        // Scanner uses other queue to send events. We must edit UI in the main queue
        DispatchQueue.main.async {
            self.logger!.bluetoothManager = nil
            self.connectionButton.setTitle("CONNECT", for: .normal)
            self.deviceName.text = "DEFAULT UART"
            
            if AppUtilities.isApplicationInactive() {
                AppUtilities.showBackgroundNotification(message: "Peripheral \(self.uartPeripheralName!) is disconnected")
            }

            self.uartPeripheralName = nil
            NotificationCenter.default.removeObserver(self, name: UIApplication.didBecomeActiveNotification, object: nil)
            NotificationCenter.default.removeObserver(self, name: UIApplication.didEnterBackgroundNotification, object: nil)
        }
        bluetoothManager = nil
    }

    //MARK: - ButtonconfigureDelegate
    func didConfigureButton(_ aButton: UIButton, withCommand aCommand: String, andIconIndex index: Int, shouldHide hide: Bool) {
        let userDefaults = UserDefaults.standard
        let buttonTag = (selectedButton?.tag)!-1
        buttonsHiddenStatus![(selectedButton?.tag)!-1] = NSNumber(value: hide as Bool)
        
        userDefaults.set(buttonsHiddenStatus, forKey: "buttonsHiddenStatus")
        if hide == true {
            selectedButton?.setImage(nil, for: .normal)
        }else{
            let image = UIImage(named: buttonIcons[index] as! String)
            selectedButton?.setImage(image, for: .normal)
        }
        
        buttonsImageNames![buttonTag] = buttonIcons[index]
        buttonsCommands![buttonTag] = aCommand
        
        userDefaults.set(buttonsImageNames, forKey: "buttonsImageNames")
        userDefaults.set(self.buttonsCommands, forKey: "buttonsCommands")
        userDefaults.synchronize()
    }
    
    //MARK: - UArtViewController Implementation
    func retrieveButtonsConfiguration() {
        let userDefaults = UserDefaults.standard
        
        if userDefaults.object(forKey: "buttonsCommands") != nil {
            //Buttons configurations already saved in NSUserDefaults
            buttonsCommands = NSMutableArray(array: userDefaults.object(forKey: "buttonsCommands") as! NSArray)
            buttonsHiddenStatus = NSMutableArray(array: userDefaults.object(forKey: "buttonsHiddenStatus") as! NSArray)
            buttonsImageNames   = NSMutableArray(array: userDefaults.object(forKey: "buttonsImageNames") as! NSArray)
             showButtonsWithSavedConfiguration()
        } else {
            //First time viewcontroller is loaded and there is no saved buttons configurations in NSUserDefaults
            //Setting up the default values for the first time
            buttonsCommands = NSMutableArray(array: ["","","","","","","","",""])
            buttonsHiddenStatus = NSMutableArray(array: [true,true,true,true,true,true,true,true,true])
            buttonsImageNames = NSMutableArray(array: ["Play","Play","Play","Play","Play","Play","Play","Play","Play"])
            userDefaults.set(buttonsCommands, forKey: "buttonsCommands")
            userDefaults.set(buttonsHiddenStatus, forKey: "buttonsHiddenStatus")
            userDefaults.set(buttonsImageNames, forKey: "buttonsImageNames")
            userDefaults.synchronize()
        }
    }
    
    func showButtonsWithSavedConfiguration() {
        for aButton: UIButton in buttons {
            if (buttonsHiddenStatus![aButton.tag - 1] as AnyObject).boolValue == true {
                aButton.backgroundColor = .nordicLightGray
                aButton.setImage(nil, for: .normal)
                aButton.isEnabled = false
            } else {
                aButton.backgroundColor = .nordicLake
                aButton.setImage(UIImage(named: buttonsImageNames![aButton.tag-1] as! String), for: .normal)
                aButton.isEnabled = true
            }
        }
    }
    
    func showPopoverOnButton() {
        let popOverViewController = storyboard?.instantiateViewController(withIdentifier: "StoryboardIDEditPopup") as! EditPopupViewController
        popOverViewController.delegate = self
        popOverViewController.isHidden = false
        popOverViewController.command = buttonsCommands![(selectedButton?.tag)!-1] as? String
        let buttonImageName = buttonsImageNames![(selectedButton?.tag)!-1]
        popOverViewController.setIconIndex(buttonIcons.index(of: buttonImageName))
        popOverViewController.modalPresentationStyle = UIModalPresentationStyle.popover
        popOverViewController.popoverPresentationController?.delegate = self
        present(popOverViewController, animated: true, completion: nil)

        popOverViewController.popoverPresentationController?.sourceView = self.view!
        popOverViewController.popoverPresentationController?.sourceRect = self.view.bounds
        popOverViewController.popoverPresentationController?.permittedArrowDirections = UIPopoverArrowDirection(rawValue: 0)
        popOverViewController.preferredContentSize = CGSize(width: 300.0, height: 300.0)
    }

    func setEditMode(mode aMode: Bool){
        editMode = aMode
        
        if aMode {
            editButton.setTitle("Done", for: .normal)
            for aButton : UIButton in buttons {
                aButton.backgroundColor = .nordicFall
                aButton.isEnabled = true
            }
        } else {
            editButton.setTitle("Edit", for: .normal)
            showButtonsWithSavedConfiguration()
        }
    }

    @objc func applicationDidEnterBackgroundCallback(){
        AppUtilities.showBackgroundNotification(message: "You are still connected to \(self.uartPeripheralName!)")
    }
    
    @objc func applicationDidBecomeActiveCallback(){
        UIApplication.shared.cancelAllLocalNotifications()
    }
    
    //MARK: - UART API
    func send(value aValue : String) {
        bluetoothManager?.send(text: aValue)
    }
}
